src/runtime-rs/crates/resource/src/volume/hugepage.rs

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{BufRead, BufReader},
+};
+
+use crate::share_fs::EPHEMERAL_PATH;
+use agent::Storage;
+use anyhow::{anyhow, Context, Ok, Result};
+use async_trait::async_trait;
+use byte_unit::Byte;
+use hypervisor::HUGETLBFS;
+use kata_sys_util::{fs::get_base_name, mount::PROC_MOUNTS_FILE};
+use kata_types::mount::KATA_EPHEMERAL_VOLUME_TYPE;
+
+use super::{Volume, BIND};
+
+type PageSize = Byte;
+type Limit = u64;
+
+const NODEV: &str = "nodev";
+
+// container hugepage
+pub(crate) struct Hugepage {
+    // storage info
+    storage: Option<Storage>,
+    // mount info
+    mount: oci::Mount,
+}
+
+// handle hugepage
+impl Hugepage {
+    pub(crate) fn new(
+        mount: &oci::Mount,
+        hugepage_limits_map: HashMap<PageSize, Limit>,
+        fs_options: Vec<String>,
+    ) -> Result<Self> {
+        // Create mount option string
+        let page_size = get_page_size(fs_options).context("failed to get page size")?;
+        let option = hugepage_limits_map
+            .get(&page_size)
+            .map(|limit| format!("pagesize={},size={}", page_size.get_bytes(), limit))
+            .context("failed to get hugepage option")?;
+        let base_name = get_base_name(mount.source.clone())?
+            .into_string()
+            .map_err(|e| anyhow!("failed to convert to string{:?}", e))?;
+        let mut mount = mount.clone();
+        // Set the mount source path to a path that resides inside the VM
+        mount.source = format!("{}{}{}", EPHEMERAL_PATH, "/", base_name);
+        // Set the mount type to "bind"
+        mount.r#type = BIND.to_string();
+
+        // Create a storage struct so that kata agent is able to create
+        // hugetlbfs backed volume inside the VM
+        let storage = Storage {
+            driver: KATA_EPHEMERAL_VOLUME_TYPE.to_string(),
+            source: NODEV.to_string(),
+            fs_type: HUGETLBFS.to_string(),
+            mount_point: mount.source.clone(),
+            options: vec![option],
+            ..Default::default()
+        };
+        Ok(Self {
+            storage: Some(storage),
+            mount,
+        })
+    }
+}
+
+#[async_trait]
+impl Volume for Hugepage {
+    fn get_volume_mount(&self) -> Result<Vec<oci::Mount>> {
+        Ok(vec![self.mount.clone()])
+    }
+
+    fn get_storage(&self) -> Result<Vec<agent::Storage>> {
+        let s = if let Some(s) = self.storage.as_ref() {
+            vec![s.clone()]
+        } else {
+            vec![]
+        };
+        Ok(s)
+    }
+
+    async fn cleanup(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub(crate) fn get_huge_page_option(m: &oci::Mount) -> Result<Option<Vec<String>>> {
+    if m.source.is_empty() {
+        return Err(anyhow!("empty mount source"));
+    }
+    let file = File::open(PROC_MOUNTS_FILE).context("failed open file")?;
+    let reader = BufReader::new(file);
+    for line in reader.lines().flatten() {
```

Comment:
Since you assume the only mount point used is `/dev/hugepages`, what if only hugetlbfs is detected here?